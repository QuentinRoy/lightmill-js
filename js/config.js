define({
    SERVER_ADDRESS: 'http://0.0.0.0:5000',
    TARGET_EXPERIMENT: 'nonmode'
});
