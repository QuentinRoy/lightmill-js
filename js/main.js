/*jslint browser:true*/
/*global requirejs, define */

requirejs.config({
    enforceDefine: true,
    paths: {
        'jquery': 'libs/jquery-2.0.0',
        'raphaeljs': 'libs/raphael-arrow-fix',
        'text': 'libs/text',
        'jstools': 'packages/jstools/src',
        'templates': '../templates',
        'state-machine': 'libs/state-machine-modified',
        'signals': 'libs/signals',
        'cookies': 'libs/cookies',
        'underscore': 'libs/underscore',
        'color': 'libs/color-0.4.4',
    },
    packages: [
        {
            name: 'sigmamenu',
            location: 'packages/sigmamenu/src/',
            main: 'sigma-menu'
        }
    ],
    shim: {
        underscore: {
            exports: '_'
        },

        color: {
            exports: 'Color'
        },
    }
});


define(['xpclient/manager', 'xpclient/tasks/task-factory'], function (XpManager, TaskFactory) {
    'use strict';
    var taskFactory = new TaskFactory(),
        manager = new XpManager(taskFactory);
    manager.start();
});