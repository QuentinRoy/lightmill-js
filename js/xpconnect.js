define(['config', 'jquery'], function (config, $) {

    function XpConnect() {
        this.serverAddress = config.SERVER_ADDRESS;
        this.targetXp = config.TARGET_EXPERIMENT;
        this.runId = null;

        this._request_run();
    }

    XpConnect.prototype = {

        _request_run: function () {
            var address = this.serverAddress + '/experiment/' + this.targetXp + '/next_run';
            $.get(address).done($.proxy(this._set_run, this)).fail(function () {
                console.log('Cannot get target run for experiment ' + this.targetXp + ' at ' + address);
            });
        },

        _set_run: function (runId) {
            this.runId = runId;
        }
    };

    return new XpConnect();

});