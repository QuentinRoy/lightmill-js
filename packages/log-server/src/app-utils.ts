import type { paths } from '@lightmill/log-api';
import type { SessionData } from 'express-session';
import { groupBy, intersection, map, pipe, uniqueBy } from 'remeda';
import type { Simplify, WritableDeep } from 'type-fest';
import {
  type ApiPath,
  type HttpMethod,
  httpStatuses,
  type HttpStatusMap,
} from './api-utils.js';
import type { RunId, Store } from './store.js';
import type {
  Handler,
  HandlerParameters,
  HandlerResult,
  RequestContent,
  ServerDescription,
} from './typed-server.js';
import { arrayify } from './utils.js';

declare module 'express-session' {
  interface SessionData {
    data: { role: 'participant' | 'host'; runs: RunId[] };
  }
}

// This needs to be a type (not an interface) so we can use it with
// typed-server... I don't know why, but it's not worth investigating.
export type ServerApi = Simplify<paths>;

export type ServerHandler<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = Handler<ServerApi, Path, Method>;

export type ServerHandlerResult<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = HandlerResult<ServerApi, Path, Method>;

export type ServerHandlerBody<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = ServerHandlerResult<Path, Method>['body'];

export type ServerHandlerIncluded<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> =
  ServerHandlerBody<Path, Method> extends infer B
    ? B extends { readonly included?: infer I }
      ? I
      : never
    : never;

export type ServerHandlerParameter<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = HandlerParameters<ServerApi, Path, Method>;

export type ServerRequestContent<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = RequestContent<ServerApi, Path, Method>;

export function getErrorResponse<
  const T extends {
    status: keyof HttpStatusMap;
    code: string;
    detail?: string;
    source?: object;
  },
>(option: T) {
  let error: WritableDeep<Omit<T, 'status'>> & {
    status: HttpStatusMap[T['status']];
  } = { ...structuredClone(option), status: httpStatuses[option.status] };
  return {
    status: option.status as T['status'],
    body: { errors: [error] },
    contentType: 'application/json' as const,
  };
}

export type SubServerDescription<K extends string> = Pick<
  ServerDescription<ServerApi>,
  Extract<keyof ServerDescription<ServerApi>, `${K}${string}`>
>;

type GetRunResourcesOptions =
  | { filter: Parameters<Store['getRuns']>[0] }
  | {
      runs: Omit<
        Awaited<ReturnType<Store['getRuns']>>[number],
        'runCreatedAt'
      >[];
    };
export async function getRunResources(
  store: Store,
  options: GetRunResourcesOptions,
) {
  const runs =
    'runs' in options ? options.runs : await store.getRuns(options.filter);
  const [experiments, lastLogs] = await Promise.all([
    store.getExperiments({
      experimentId: pipe(
        runs,
        uniqueBy((run) => run.experimentId),
        map((run) => run.experimentId),
      ),
    }),
    store.getLastLogs({ runId: runs.map((run) => run.runId) }),
  ]);
  const groupedLastLogs = groupBy(lastLogs, (log) => log.runId);

  return {
    runs: runs.map((run) => {
      const runLastLogs = groupedLastLogs[run.runId] ?? [];
      return {
        id: run.runId,
        type: 'runs' as const,
        attributes: {
          status: run.runStatus,
          name: run.runName,
          lastLogNumber: Math.max(0, ...runLastLogs.map((l) => l.number)),
        },
        relationships: {
          lastLogs: {
            data: runLastLogs.map((log) => ({
              id: log.logId,
              type: 'logs' as const,
            })),
          },
          experiment: {
            data: { id: run.experimentId, type: 'experiments' as const },
          },
        },
      };
    }),
    experiments: experiments.map((experiment) => ({
      id: experiment.experimentId,
      type: 'experiments' as const,
      attributes: { name: experiment.experimentName },
    })),
    lastLogs: lastLogs.map((log) => ({
      id: log.logId,
      type: 'logs' as const,
      attributes: { number: log.number, logType: log.type, values: log.values },
      relationships: {
        run: { data: { id: log.runId, type: 'runs' as const } },
      },
    })),
  };
}

export function getAllowedAndFilteredRunIds(
  sessionData: SessionData['data'] | undefined,
  queryFilter: undefined | string | string[],
) {
  if (sessionData == null) {
    return [];
  }
  if (sessionData.role === 'host') {
    return queryFilter;
  }
  if (queryFilter == null) {
    return sessionData.runs;
  }
  return intersection(sessionData.runs, arrayify(queryFilter, true));
}
