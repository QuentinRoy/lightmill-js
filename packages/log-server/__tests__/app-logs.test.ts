/* eslint-disable no-empty-pattern */

import { describe, vi, test as vitestTest } from 'vitest';
import { apiMediaType } from '../src/app-utils.ts';
import { StoreError } from '../src/store.js';
import {
  apiContentTypeRegExp,
  createFixtureWithRuns,
  type FixtureWithRuns,
} from './test-utils.js';

const it = vitestTest.extend<{
  participantFixture: FixtureWithRuns<'participant'>;
  hostFixture: FixtureWithRuns<'host'>;
}>({
  participantFixture: async ({}, use) => {
    const fixture = await createFixtureWithRuns({ role: 'participant' });
    fixture.setRuns([
      {
        runId: 'run-id',
        runName: 'run-name',
        experimentId: fixture.experiment,
        runStatus: 'running' as const,
        runCreatedAt: vi.getMockedSystemTime() ?? new Date(),
      },
    ]);
    vi.useFakeTimers({ toFake: ['Date'] });
    vi.setSystemTime(100100);
    await use(fixture);
    vi.useRealTimers();
  },
  hostFixture: async ({}, use) => {
    const fixture = await createFixtureWithRuns({ role: 'host' });
    fixture.setRuns([
      {
        runId: 'run-id',
        runName: 'run-name',
        experimentId: fixture.experiment,
        runStatus: 'running' as const,
        runCreatedAt: vi.getMockedSystemTime() ?? new Date(),
      },
    ]);
    vi.useFakeTimers({ toFake: ['Date'] });
    vi.setSystemTime(100100);
    await use(fixture);
    vi.useRealTimers();
  },
});

describe('LogServer: post /logs', () => {
  it('adds a log', async ({
    expect,
    participantFixture: { api, store, getRuns },
  }) => {
    const runId = getRuns()[0]!.runId;
    const response = await api
      .post('/logs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 1, logType: 'test', values: { x: 'x' } },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201)
      .expect('location', 'http://lightmill-test.com/logs/l1');

    expect(response.body).toEqual({
      data: { type: 'logs', id: expect.any(String) },
    });
    expect(store.addLogs).toHaveBeenCalledWith(runId, [
      { type: 'test', values: { x: 'x' }, number: 1 },
    ]);
  });

  it('refuses to add logs if the client does not have access to the run', async ({
    expect,
    participantFixture: {
      store,
      sessionStore,
      api,
      setRuns,
      getRuns,
      experiment,
    },
  }) => {
    setRuns([
      ...getRuns(),
      {
        runId: 'not-my-run',
        runName: 'not-my-run',
        experimentId: experiment,
        runStatus: 'running' as const,
        runCreatedAt: vi.getMockedSystemTime() ?? new Date(),
      },
    ]);
    sessionStore.mockGetData({ role: 'participant', runs: ['my-run'] });
    await api
      .post('/logs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 1, logType: 'test', values: { x: 'x' } },
          relationships: { run: { data: { type: 'runs', id: 'not-my-run' } } },
        },
      })
      .expect(403, {
        errors: [
          {
            status: 'Forbidden',
            code: 'RUN_NOT_FOUND',
            detail: "Run 'not-my-run' not found",
          },
        ],
      })
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.addLogs).not.toHaveBeenCalled();
  });

  it('allows hosts to add logs to any run', async ({
    hostFixture: { api, sessionStore, getRuns, setRuns, experiment },
  }) => {
    setRuns([
      ...getRuns(),
      {
        runId: 'not-my-run',
        runName: 'not-my-run',
        experimentId: experiment,
        runStatus: 'running' as const,
        runCreatedAt: vi.getMockedSystemTime() ?? new Date(),
      },
    ]);
    sessionStore.mockGetData({ role: 'host', runs: ['my-run'] });
    await api
      .post('/logs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 1, logType: 'test', values: { x: 'x' } },
          relationships: { run: { data: { type: 'runs', id: 'not-my-run' } } },
        },
      })
      .expect(201);
  });

  it('refuses to add logs to a run that does not exist', async ({
    expect,
    participantFixture: { store, api },
  }) => {
    await api
      .post('/logs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 1, logType: 'test', values: { x: 'x' } },
          relationships: {
            run: { data: { type: 'runs', id: 'does-not-exist' } },
          },
        },
      })
      .expect(403, {
        errors: [
          {
            status: 'Forbidden',
            code: 'RUN_NOT_FOUND',
            detail: "Run 'does-not-exist' not found",
          },
        ],
      })
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.addLogs).not.toHaveBeenCalled();
  });

  it('refuses to add logs if their number is already in used', async ({
    participantFixture: { store, api, getRuns },
  }) => {
    const runId = getRuns()[0]!.runId;
    store.addLogs.mockImplementation(async () => {
      throw new StoreError(
        'Log number 2 is already used',
        StoreError.LOG_NUMBER_EXISTS_IN_SEQUENCE,
      );
    });
    await api
      .post('/logs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 2, logType: 'test-log', values: {} },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(409, {
        errors: [
          {
            status: 'Conflict',
            code: 'LOG_NUMBER_EXISTS',
            detail:
              "Cannot add log to run 'run-id', log number 2 already exists",
          },
        ],
      })
      .expect('Content-Type', apiContentTypeRegExp);
  });
});

describe('LogServer: get /logs', () => {
  it('returns logs as csv by default', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    let result = await api
      .get('/logs')
      .expect(200)
      .expect('Content-Type', /^text\/csv/);
    expect(store.getLogs.mock.calls).toMatchSnapshot();
    expect(result.text).toMatchSnapshot();
  });

  it('returns logs as json if json is the first supported format in the Accept header', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    let result = await api
      .get('/logs')
      .set('Accept', apiMediaType)
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchSnapshot();
    expect(result.body).toMatchSnapshot();
  });

  it('returns logs as csv if csv is the first supported format in the Accept header', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    let result = await api.get('/logs').set('Accept', 'text/csv').expect(200);
    expect(store.getLogs.mock.calls).toMatchSnapshot();
    expect(result.text).toMatchSnapshot();
  });

  it('returns a 400 error if the Accept header is not supported', async ({
    hostFixture: { api },
  }) => {
    await api
      .get('/logs')
      .set('Accept', 'application/xml,application/pdf,text/html')
      .expect(400, {
        errors: [
          {
            code: 'HEADERS_VALIDATION',
            detail:
              'must be equal to one of the allowed values: application/vnd.api+json, text/csv',
            source: { header: 'accept' },
            status: 'Bad Request',
          },
        ],
      })
      .expect('Content-Type', apiContentTypeRegExp);
  });

  it('returns only logs a participant has access to', async ({
    expect,
    participantFixture: { api, store, sessionStore },
  }) => {
    const runBase = {
      experimentId: 'exp',
      runCreatedAt: new Date('2023-01-01'),
      runName: 'run',
      runStatus: 'completed' as const,
    };
    store.getRuns.mockImplementation(async () => [
      { runId: 'that-run', ...runBase },
      { runId: 'another-run', ...runBase },
    ]);
    sessionStore.mockGetData({ role: 'participant', runs: ['that-run'] });

    let result = await api
      .set('Accept', apiMediaType)
      .get('/logs')
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchSnapshot();
    expect(result.body).toMatchSnapshot();
  });

  it('filters logs by type', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    const response = await api
      .get('/logs')
      .set('Accept', apiMediaType)
      .query({ 'filter[logType]': 'log-type' })
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "experimentId": undefined,
            "logType": "log-type",
            "runId": undefined,
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
    expect(response.body).toMatchSnapshot();
  });

  it('filters logs by experiment', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    const response = await api
      .get('/logs')
      .set('Accept', apiMediaType)
      .query({ 'filter[experiment.id]': 'experiment-id' })
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "experimentId": "experiment-id",
            "logType": undefined,
            "runId": undefined,
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
    expect(response.body).toMatchSnapshot();
  });

  it('filters logs by run', async ({ expect, hostFixture: { api, store } }) => {
    const response = await api
      .get('/logs')
      .set('Accept', apiMediaType)
      .query({ 'filter[run.id]': 'run-id' })
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "experimentId": undefined,
            "logType": undefined,
            "runId": "run-id",
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
    expect(response.body).toMatchSnapshot();
  });

  it('filters logs by run, type, and experiment', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    const response = await api
      .get('/logs')
      .set('Accept', apiMediaType)
      .query({
        'filter[logType]': 'log-type',
        'filter[experiment.id]': 'experiment-id',
        'filter[run.id]': 'run-id',
      })
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "experimentId": "experiment-id",
            "logType": "log-type",
            "runId": "run-id",
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
    expect(response.body).toMatchSnapshot();
  });
});

describe('LogServer: get /logs/{id}', () => {
  it("returns a 404 error if the log is not part of one of the participant's runs", async ({
    expect,
    participantFixture: { api, store },
  }) => {
    store.getLogs.mockImplementation(async function* () {});
    await api
      .get('/logs/not-mine')
      .expect(404, {
        errors: [
          {
            status: 'Not Found',
            code: 'LOG_NOT_FOUND',
            detail: "Log 'not-mine' not found",
          },
        ],
      })
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "logId": "not-mine",
            "runId": [
              "run-id",
            ],
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
  });

  it('returns a 404 error if the log does not exist', async ({
    expect,
    hostFixture: { api, store },
  }) => {
    store.getLogs.mockImplementation(async function* () {});
    await api
      .get('/logs/does-not-exist')
      .expect(404, {
        errors: [
          {
            status: 'Not Found',
            code: 'LOG_NOT_FOUND',
            detail: "Log 'does-not-exist' not found",
          },
        ],
      })
      .expect('Content-Type', apiContentTypeRegExp);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "logId": "does-not-exist",
            "runId": undefined,
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
  });

  it('returns the log', async ({
    expect,
    participantFixture: { api, store },
  }) => {
    store.getLogs.mockImplementation(async function* () {
      yield {
        number: 5,
        logId: 'log-id',
        type: 'log-type',
        experimentId: 'experiment-id',
        experimentName: 'experiment-name',
        runName: 'run-name',
        runStatus: 'running' as const,
        runId: 'run-id',
        createdAt: new Date('2024-01-01T00:00:00Z'),
        values: { x: 'x' },
      };
    });
    let result = await api.get('/logs/log-id').expect(200);
    expect(store.getLogs.mock.calls).toMatchInlineSnapshot(`
      [
        [
          {
            "logId": "log-id",
            "runId": [
              "run-id",
            ],
            "runStatus": "-canceled",
          },
        ],
      ]
    `);
    expect(result.body).toMatchInlineSnapshot(`
      {
        "data": {
          "attributes": {
            "logType": "log-type",
            "number": 5,
            "values": {
              "x": "x",
            },
          },
          "id": "log-id",
          "relationships": {
            "run": {
              "data": {
                "id": "run-id",
                "type": "runs",
              },
            },
          },
          "type": "logs",
        },
      }
    `);
  });
});
