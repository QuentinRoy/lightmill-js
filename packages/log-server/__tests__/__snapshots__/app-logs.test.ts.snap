// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogServer: get /logs > filters logs by experiment 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`LogServer: get /logs > filters logs by run 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`LogServer: get /logs > filters logs by run, type, and experiment 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`LogServer: get /logs > filters logs by type 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`LogServer: get /logs > returns logs as csv if csv is the first supported format in the Accept header 1`] = `
[
  [
    {
      "experimentId": undefined,
      "logType": undefined,
      "runId": undefined,
      "runStatus": "-canceled",
    },
  ],
]
`;

exports[`LogServer: get /logs > returns logs as csv if csv is the first supported format in the Accept header 2`] = `
"type,experiment_name,run_name,run_status,mock_col_1,mock_col_2,mock_col_3
getLogs:type-1,getLogs:experimentName-1,getLogs:runName-1,running,log1-mock-value1,log1-mock-value2,
getLogs:type-2,getLogs:experimentName-2,getLogs:runName-2,completed,log2-mock-value1,log2-mock-value2,log2-mock-value3
"
`;

exports[`LogServer: get /logs > returns logs as json by default 1`] = `
[
  [
    {
      "experimentId": undefined,
      "logType": undefined,
      "runId": undefined,
      "runStatus": [
        "-canceled",
      ],
    },
  ],
]
`;

exports[`LogServer: get /logs > returns logs as json by default 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`LogServer: get /logs > returns logs as json if json is the first supported format in the Accept header 1`] = `
[
  [
    {
      "experimentId": undefined,
      "logType": undefined,
      "runId": undefined,
      "runStatus": [
        "-canceled",
      ],
    },
  ],
]
`;

exports[`LogServer: get /logs > returns logs as json if json is the first supported format in the Accept header 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`LogServer: get /logs > returns only logs a participant has access to 1`] = `
[
  [
    {
      "experimentId": undefined,
      "logType": undefined,
      "runId": [
        "that-run",
      ],
      "runStatus": [
        "-canceled",
      ],
    },
  ],
]
`;

exports[`LogServer: get /logs > returns only logs a participant has access to 2`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "getLogs:type-1",
        "number": 1,
        "values": {
          "mock-col1": "log1-mock-value1",
          "mock-col2": "log1-mock-value2",
        },
      },
      "id": "getLogs:id-1",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-1",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
    {
      "attributes": {
        "logType": "getLogs:type-2",
        "number": 2,
        "values": {
          "mock-col1": "log2-mock-value1",
          "mock-col2": "log2-mock-value2",
          "mock-col3": "log2-mock-value3",
        },
      },
      "id": "getLogs:id-2",
      "relationships": {
        "run": {
          "data": {
            "id": "getLogs:run-id-2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;
